harness/determined/experimental/core_v2/_core_v2.py

```diff
@@ -149,7 +152,14 @@ def _init_context(
     # Construct the config.
     defaulted_config = defaults or DefaultConfig()
     unmanaged_config = unmanaged or UnmanagedConfig()
-    checkpoint_storage = checkpoint_storage or defaulted_config.checkpoint_storage
+    checkpoint_storage = (
+        checkpoint_storage
+        or defaulted_config.checkpoint_storage
+        or str(pathlib.Path(appdirs.user_data_dir("determined")) / "checkpoints")
+    )
+
+    if type(checkpoint_storage) == str:
```

Comment:
it's possible someone passes in `checkpoint_storage` or configures it in the `DefaultConfig` as a string, right? so this would effectively override the passed in value?

this snippet might be cleaner as:

```python3
    checkpoint_storage = checkpoint_storage or defaulted_config.checkpoint_storage

    if checkpoint_storage is None:
        checkpoint_storage = {
            "type": "directory",
            "container_path": str(
                pathlib.Path(appdirs.user_data_dir("determined")) / "checkpoints"
            ),
        }

```

Reply:
I might be wrong, but `A or B` only equals to B when A is False/None/empty, right?
Also the `checkpoint_storage` parameter and `defaulted_config.checkpoint_storage` are both string, so it needs to be wrapped too before putting into the config

Reply:
i meant that it's possible for checkpoint storage to be passed in as a string but it might not be `directory` type, i think?

looks like we have some logic in `_core_context_v2.py` that builds a storage manager from `checkpoint_storage`, and we accept string types that can be gs/s3 types: 
```
def _get_storage_manager(
    checkpoint_storage: Optional[Union[str, Dict[str, Any]]]
) -> Optional[storage.StorageManager]:
    if checkpoint_storage is None:
        return None
    if isinstance(checkpoint_storage, str):
        return storage.from_string(checkpoint_storage)
```

maybe this codepath doesn't get called? i haven't read the code super thoroughly. 

i thought what we were trying to do here is set `checkpoint_storage` to `{"type": "directory", "container_path": checkpoint_storage}` only in the default case when it wasn't passed in. but if it is passed in, either directly as a parameter or in `defaulted_config`, then i'm not sure we can assume it's type=directory and the `container_path`. 

at least that's what i was thinking when i commented. if it should always be directory/container_path, then ignore me 🙃

Reply:
Got it, thanks for the explanation! 
I noticed that we used to only parse non-cloud storage to `shared_fs`, both in [storage.from_string](https://github.com/determined-ai/determined/blob/a4980083505b60f7b64c8fd01e1de7734e1bda0d/harness/determined/common/storage/base.py#L223) and [_shortcut_to_config](https://github.com/determined-ai/determined/blob/a4980083505b60f7b64c8fd01e1de7734e1bda0d/harness/determined/common/storage/shared.py#L101), is there a way to distinguish between type `shared_fs` and `directory`? 
Upon testing, I don't think there is going to be any difference in terms of WebUI display.

Reply:
hm, yeah there's no way to distinguish between "file://" for shared fs vs directory if passed in. i think we can determine it by on vs off cluster. on-cluster should parse as shared fs, off as directory.

basically something like:

if checkpoint_storage is a string type:
    if on_cluster:
        parse as shared_fs
    else:
        parse as directory
if checkpoint_storage is None:
    path = pathlib.Path(appdirs.user_data_dir("determined")) / "checkpoints"
    if on_cluster:
        shared_fs(path)
    else:
        directory(path)

Reply:
Thanks for the feedback! 
So for on-cluster experiments, setting checkpoint storage is handled inside [_make_v2_context](https://github.com/determined-ai/determined/blob/56a8784ff585777b3813a5e43d37c856ae9309f6/harness/determined/experimental/core_v2/_core_context_v2.py#L76), and we probably don't need to change the logic there. 
Then for off-cluster experiments, we can explicitly set the checkpoint storage to an [object](https://github.com/determined-ai/determined/blob/56a8784ff585777b3813a5e43d37c856ae9309f6/harness/determined/experimental/core_v2/_core_v2.py#L177), so it will be the same type either in config and in context.